docs/proposals/provisioning-proposal.md

```diff
@@ -0,0 +1,112 @@
+# Cluster-API Provisioning Mechanism Consolidation Proposal
+
+## Abstract
+
+Cluster-API requires implementers to implement provisioning mechanisms (bootstrap scripts and logic for running those scripts) themselves. However, the mechanism, including bootstrap scripts, is same or very similar for all cloud providers. As this is a time consuming task it could be very useful to provide a reusable provisioning mechanism in the upstream Cluster-API.
+
+## Proposal
+
+The upstream Cluster-API doesn’t have any bootstrap scripts neither any wrapper around kubeadm for handling cluster operations (initialization and join). This means implementers have to manually write bootstrap scripts used by specific Cluster-API provider and manually handle cluster initialization and node joining. For most cloud providers both bootstrap scripts and steps for cluster initialization/joining are the same, except for provider specifics that could be provided via template variables.
+
+The bootstrap scripts usually do the following tasks:
+
+* Install dependencies required by Kubernetes
+    * Includes dependencies such as socat, ebtables, which are same for all providers.
+* Install container runtime
+    * Steps for installing container runtime such as Docker or containerd are same for all providers.
```

Comment:
@roberthbailey: For a specific runtime. But different providers may choose not to implement support for all runtimes. If this code is standardized, does that imply a burden of supporting all runtimes that are upstream on a provider?

@alvaroaleman: I think we will provide just one script for one runtime that works, if ppl want something else they can choose not to use that script